File: project-root/asm-msa-customer/src/main/java/com/company/asm/customer/service/CustomerService.java
package com.company.asm.customer.service;

import com.company.asm.customer.api.model.CustomerInput;
import com.company.asm.customer.api.model.CustomerOutput;
import com.company.asm.customer.api.model.CustomerPatchInput;

import java.util.List;
import java.util.UUID;

public interface CustomerService {
    CustomerOutput createCustomer(CustomerInput customerInput);
    CustomerOutput getCustomerById(UUID customerId);
    List<CustomerOutput> getAllCustomers();
    CustomerOutput updateCustomer(UUID customerId, CustomerInput customerInput);
    CustomerOutput patchCustomer(UUID customerId, CustomerPatchInput customerPatchInput);
    void deleteCustomer(UUID customerId);
}

File: project-root/app-spa-portal/src/app/features/customer/services/customer/customer.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { SERVICE_URL_ENDPOINTS } from '../../../../core/config/service-url-endpoints';
import { ICustomer } from '../../models/customer.model';
import { IMessageResponse } from '../../../../shared/models/message-response.model';

@Injectable({
  providedIn: 'root'
})
export class CustomerService {
  constructor(private readonly http: HttpClient) { }

  getAllCustomers() {
    return this.http.get<ICustomer[]>(SERVICE_URL_ENDPOINTS.CUSTOMERS_ALL);
  }

  getCustomer(idCustomer: string) {
    return this.http.get<ICustomer>(`${SERVICE_URL_ENDPOINTS.CUSTOMER}/${idCustomer}`);
  }

  updateCustomer(idCustomer: string, data: ICustomer) {
    return this.http.put<ICustomer>(`${SERVICE_URL_ENDPOINTS.CUSTOMER}/${idCustomer}`, data);
  }

  updateCustomerPartial(idCustomer: string, data: ICustomer) {
    return this.http.patch<ICustomer>(`${SERVICE_URL_ENDPOINTS.CUSTOMER}/${idCustomer}`, data);
  }

  deleteCustomer(idCustomer: string) {
    return this.http.delete<IMessageResponse>(`${SERVICE_URL_ENDPOINTS.CUSTOMER}/${idCustomer}`);
  }

  createCustomer(data: ICustomer) {
    return this.http.post<ICustomer>(SERVICE_URL_ENDPOINTS.CUSTOMER_CREATE, data);
  }

}
